### PR TITLE
Query builder

### PR DESCRIPTION
**Pull request by dantleech**
This PR implements the query builder. The code has been ported from the earlier, abandoned, metadata PR.

Example using a join:

``` php
    $builder = $this->getDocumentManager()->createQueryBuilder();
    $builder->setLocale('en');
    $builder->from('i')->joinInner()
        ->left()->document('full', 'f')->end()
        ->right()->document('issue', 'i')->end()
        ->condition()->equi('f.status', 'i.status')
        ->where()->eq()->field('i.status')->literal('closed');
    $results = $builder->getQuery()->execute();
```

It is uses the PHPCR-ODM query builder and introduces a "converter" to translate integrate the query builder with the Sulu Document Manager.

Note that this PR does not include any specific Sulu Content functionality. That will be introduced in a follow up PR on the sulu-io/sulu repository.

This PR depends on the imminent PHPCR-ODM 1.3 release.

### CODE REVIEW

File: lib/Metadata.php [hunk 1]
```diff
@@ -82,6 +82,25 @@ public function getFieldMappings()
...
+     * Return the mapping for the named field.
+     *
+     * @see Metadata::addFieldMapping
+     *
+     * @return array
+     */
+    public function getFieldMapping($name)
+    {
+        if (!isset($this->fieldMappings[$name])) {
+            throw new \InvalidArgumentException(sprintf(
```
Comment (dantleech) [inline lib/Metadata.php]: Test this exception

File: lib/Query/QueryBuilderConverter.php [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+use Sulu\Component\DocumentManager\PropertyEncoder;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class which converts a Builder tree to a PHPCR Query.
+ */
+class QueryBuilderConverter extends ConverterPhpcr
+{
+    /**
+     * j.
```
Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php]: ?

File: lib/Query/QueryBuilderConverter.php [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            QBConstants::NT_WHERE,
+            QBConstants::NT_ORDER_BY,
+        ];
+
+        foreach ($dispatches as $dispatchType) {
+            $this->dispatchMany($builder->getChildrenOfType($dispatchType));
+        }
+
+        if (count($this->sourceDocumentNodes) > 1 && null === $builder->getPrimaryAlias()) {
+            throw new \InvalidArgumentException(
```
Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php]: Tested?

Comment (dantleech) [inline lib/Query/QueryBuilderConverter.php]: It is now.

**Comment by dantleech** [comment]
@danrot @chirimoya @wachterjohannes this is ready, but we need to wait for the 1.3 release of PHPCR-ODM. Feel free to review.

File: lib/DocumentStrategyInterface.php [hunk 4]
```diff
@@ -48,4 +50,48 @@ public function createNodeForDocument($document, NodeInterface $parentNode, $nam
...
     public function resolveMetadataForNode(NodeInterface $node);
+
+    /**
+     * Return the primary node type from the given document class FQN.
+     *
+     * For example:
+     *
+     * ````
+     * return 'nt:unstructured';
+     * ````
```
Comment (danrot) [inline lib/DocumentStrategyInterface.php]: Shouldn't there only be 3 ticks?

File: lib/DocumentStrategyInterface.php [hunk 5]
```diff
@@ -48,4 +50,48 @@ public function createNodeForDocument($document, NodeInterface $parentNode, $nam
...
+    public function getPrimaryNodeType($classFqn);
+
+    /**
+     * Create a source constraint for a source document. That is return the
+     * constraint that should be used to return only documents of the given
+     * class FQN.
+     *
+     * For example:
+     *
+     * ````
```
Comment (danrot) [inline lib/DocumentStrategyInterface.php]: same as above

File: tests/Unit/MetadataTest.php [hunk 6]
```diff
@@ -40,4 +40,31 @@ public function testReflectionClass()
...
         $this->assertNotSame($reflection, $this->metadata->getReflectionClass());
     }
+
+    /**
+     * It should return a specific field mapping.
+     */
+    public function testGetFieldMapping()
+    {
+        $this->metadata->addFieldMapping('hello', ['property' => 'foo']);
+        $this->assertContains([
```
Comment (danrot) [inline tests/Unit/MetadataTest.php]: I would put the `[` onto the next line.

File: lib/DocumentStrategyInterface.php (hunk 4 above)
Comment (dantleech) [inline lib/DocumentStrategyInterface.php]: I think your right, I always use 4 on github and it never complained :)

**Comment by danrot** [comment]
@dantleech There are still some open comments left.

**Comment by dantleech** [comment]
Fixed.

**Pull request closed without merging**